From 71c3ce343a62fd81fe9945ed73a7083d79596aa0 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Thu, 26 Aug 2010 16:47:41 -0600
Subject: [PATCH] Refactor class MyWatchesTable out of MyWatchedProjectsScreen

Refactor out the table class from the file
MyWatchedProjectsScreen into its own file: MyWatchesTable.
This makes the coupling between the table and the panel more
clear and isolates the internal table logic better.

Change-Id: Ie8a3ae1f3986ec533316d62ab8b4796431688de5
---
 .../client/account/MyWatchedProjectsScreen.java    |  293 +++++---------------
 .../gerrit/client/account/MyWatchesTable.java      |  178 ++++++++++++
 2 files changed, 249 insertions(+), 222 deletions(-)
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java
index 18b192e..4433c05 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java
@@ -17,13 +17,9 @@ package com.google.gerrit.client.account;
 import com.google.gerrit.client.Gerrit;
 import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.client.rpc.ScreenLoadCallback;
-import com.google.gerrit.client.ui.FancyFlexTable;
 import com.google.gerrit.client.ui.HintTextBox;
-import com.google.gerrit.client.ui.ProjectLink;
 import com.google.gerrit.client.ui.ProjectNameSuggestOracle;
 import com.google.gerrit.common.data.AccountProjectWatchInfo;
-import com.google.gerrit.reviewdb.AccountProjectWatch;
-import com.google.gerrit.reviewdb.Change.Status;
 import com.google.gwt.event.dom.client.ClickEvent;
 import com.google.gwt.event.dom.client.ClickHandler;
 import com.google.gwt.event.dom.client.KeyCodes;
@@ -31,29 +27,21 @@ import com.google.gwt.event.dom.client.KeyPressEvent;
 import com.google.gwt.event.dom.client.KeyPressHandler;
 import com.google.gwt.event.logical.shared.SelectionEvent;
 import com.google.gwt.event.logical.shared.SelectionHandler;
-import com.google.gwt.user.client.DOM;
 import com.google.gwt.user.client.ui.Button;
-import com.google.gwt.user.client.ui.CheckBox;
 import com.google.gwt.user.client.ui.FlowPanel;
 import com.google.gwt.user.client.ui.Grid;
-import com.google.gwt.user.client.ui.Label;
 import com.google.gwt.user.client.ui.SuggestBox;
-import com.google.gwt.user.client.ui.FlexTable.FlexCellFormatter;
 import com.google.gwt.user.client.ui.HTMLTable.CellFormatter;
 import com.google.gwt.user.client.ui.SuggestOracle.Suggestion;
-import com.google.gwtjsonrpc.client.VoidResult;
 
-import java.util.HashSet;
 import java.util.List;
-import java.util.Set;
 
 public class MyWatchedProjectsScreen extends SettingsScreen {
-  private WatchTable watches;
-
   private Button addNew;
   private HintTextBox nameBox;
   private SuggestBox nameTxt;
   private HintTextBox filterTxt;
+  private MyWatchesTable watchesTab;
   private Button delSel;
   private boolean submitOnSelection;
 
@@ -61,88 +49,91 @@ public class MyWatchedProjectsScreen extends SettingsScreen {
   protected void onInitUI() {
     super.onInitUI();
 
-    {
-      nameBox = new HintTextBox();
-      nameTxt = new SuggestBox(new ProjectNameSuggestOracle(), nameBox);
-      nameBox.setVisibleLength(50);
-      nameBox.setHintText(Util.C.defaultProjectName());
-      nameBox.addKeyPressHandler(new KeyPressHandler() {
-        @Override
-        public void onKeyPress(KeyPressEvent event) {
-          submitOnSelection = false;
+    createWidgets();
 
-          if (event.getCharCode() == KeyCodes.KEY_ENTER) {
-            if (nameTxt.isSuggestionListShowing()) {
-              submitOnSelection = true;
-            } else {
-              doAddNew();
-            }
-          }
-        }
-      });
-      nameTxt.addSelectionHandler(new SelectionHandler<Suggestion>() {
-        @Override
-        public void onSelection(SelectionEvent<Suggestion> event) {
-          if (submitOnSelection) {
-            submitOnSelection = false;
-            doAddNew();
-          }
-        }
-      });
+    final Grid grid = new Grid(2, 2);
+    grid.setStyleName(Gerrit.RESOURCES.css().infoBlock());
+    grid.setText(0, 0, Util.C.watchedProjectName());
+    grid.setWidget(0, 1, nameTxt);
+
+    grid.setText(1, 0, Util.C.watchedProjectFilter());
+    grid.setWidget(1, 1, filterTxt);
+
+    final CellFormatter fmt = grid.getCellFormatter();
+    fmt.addStyleName(0, 0, Gerrit.RESOURCES.css().topmost());
+    fmt.addStyleName(0, 1, Gerrit.RESOURCES.css().topmost());
+    fmt.addStyleName(0, 0, Gerrit.RESOURCES.css().header());
+    fmt.addStyleName(1, 0, Gerrit.RESOURCES.css().header());
+    fmt.addStyleName(1, 0, Gerrit.RESOURCES.css().bottomheader());
+
+    final FlowPanel fp = new FlowPanel();
+    fp.setStyleName(Gerrit.RESOURCES.css().addWatchPanel());
+    fp.add(grid);
+    fp.add(addNew);
+    add(fp);
 
-      filterTxt = new HintTextBox();
-      filterTxt.setVisibleLength(50);
-      filterTxt.setHintText(Util.C.defaultFilter());
-      filterTxt.addKeyPressHandler(new KeyPressHandler() {
-        @Override
-        public void onKeyPress(KeyPressEvent event) {
-          if (event.getCharCode() == KeyCodes.KEY_ENTER) {
+    add(watchesTab);
+    add(delSel);
+  }
+
+  protected void createWidgets() {
+    nameBox = new HintTextBox();
+    nameTxt = new SuggestBox(new ProjectNameSuggestOracle(), nameBox);
+    nameBox.setVisibleLength(50);
+    nameBox.setHintText(Util.C.defaultProjectName());
+    nameBox.addKeyPressHandler(new KeyPressHandler() {
+      @Override
+      public void onKeyPress(KeyPressEvent event) {
+        submitOnSelection = false;
+
+        if (event.getCharCode() == KeyCodes.KEY_ENTER) {
+          if (nameTxt.isSuggestionListShowing()) {
+            submitOnSelection = true;
+          } else {
             doAddNew();
           }
         }
-      });
-
-      addNew = new Button(Util.C.buttonWatchProject());
-      addNew.addClickHandler(new ClickHandler() {
-        @Override
-        public void onClick(final ClickEvent event) {
+      }
+    });
+    nameTxt.addSelectionHandler(new SelectionHandler<Suggestion>() {
+      @Override
+      public void onSelection(SelectionEvent<Suggestion> event) {
+        if (submitOnSelection) {
+          submitOnSelection = false;
           doAddNew();
         }
-      });
-
-      final Grid grid = new Grid(2, 2);
-      grid.setStyleName(Gerrit.RESOURCES.css().infoBlock());
-      grid.setText(0, 0, Util.C.watchedProjectName());
-      grid.setWidget(0, 1, nameTxt);
-
-      grid.setText(1, 0, Util.C.watchedProjectFilter());
-      grid.setWidget(1, 1, filterTxt);
+      }
+    });
 
-      final CellFormatter fmt = grid.getCellFormatter();
-      fmt.addStyleName(0, 0, Gerrit.RESOURCES.css().topmost());
-      fmt.addStyleName(0, 1, Gerrit.RESOURCES.css().topmost());
-      fmt.addStyleName(0, 0, Gerrit.RESOURCES.css().header());
-      fmt.addStyleName(1, 0, Gerrit.RESOURCES.css().header());
-      fmt.addStyleName(1, 0, Gerrit.RESOURCES.css().bottomheader());
+    filterTxt = new HintTextBox();
+    filterTxt.setVisibleLength(50);
+    filterTxt.setHintText(Util.C.defaultFilter());
+    filterTxt.addKeyPressHandler(new KeyPressHandler() {
+      @Override
+      public void onKeyPress(KeyPressEvent event) {
+        if (event.getCharCode() == KeyCodes.KEY_ENTER) {
+          doAddNew();
+        }
+      }
+    });
 
-      final FlowPanel fp = new FlowPanel();
-      fp.setStyleName(Gerrit.RESOURCES.css().addWatchPanel());
-      fp.add(grid);
-      fp.add(addNew);
-      add(fp);
-    }
+    addNew = new Button(Util.C.buttonWatchProject());
+    addNew.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        doAddNew();
+      }
+    });
 
-    watches = new WatchTable();
-    add(watches);
+    watchesTab = new MyWatchesTable();
 
     delSel = new Button(Util.C.buttonDeleteSshKey());
     delSel.addClickHandler(new ClickHandler() {
       @Override
       public void onClick(final ClickEvent event) {
-        watches.deleteChecked();
+        watchesTab.deleteChecked();
       }
     });
-    add(delSel);
   }
 
   @Override
@@ -175,7 +166,7 @@ public class MyWatchedProjectsScreen extends SettingsScreen {
             filterTxt.setEnabled(true);
 
             nameTxt.setText("");
-            watches.insertWatch(result);
+            watchesTab.insertWatch(result);
           }
 
           @Override
@@ -194,150 +185,8 @@ public class MyWatchedProjectsScreen extends SettingsScreen {
         new ScreenLoadCallback<List<AccountProjectWatchInfo>>(this) {
       @Override
       public void preDisplay(final List<AccountProjectWatchInfo> result) {
-        watches.display(result);
+        watchesTab.display(result);
       }
     });
   }
-
-  private class WatchTable extends FancyFlexTable<AccountProjectWatchInfo> {
-    WatchTable() {
-      table.setWidth("");
-      table.insertRow(1);
-      table.setText(0, 2, Util.C.watchedProjectName());
-      table.setText(0, 3, Util.C.watchedProjectColumnEmailNotifications());
-
-      final FlexCellFormatter fmt = table.getFlexCellFormatter();
-      fmt.addStyleName(0, 1, Gerrit.RESOURCES.css().iconHeader());
-      fmt.addStyleName(0, 2, Gerrit.RESOURCES.css().dataHeader());
-      fmt.addStyleName(0, 3, Gerrit.RESOURCES.css().dataHeader());
-      fmt.setRowSpan(0, 0, 2);
-      fmt.setRowSpan(0, 1, 2);
-      fmt.setRowSpan(0, 2, 2);
-      DOM.setElementProperty(fmt.getElement(0, 3), "align", "center");
-
-      fmt.setColSpan(0, 3, 3);
-      table.setText(1, 0, Util.C.watchedProjectColumnNewChanges());
-      table.setText(1, 1, Util.C.watchedProjectColumnAllComments());
-      table.setText(1, 2, Util.C.watchedProjectColumnSubmittedChanges());
-      fmt.addStyleName(1, 0, Gerrit.RESOURCES.css().dataHeader());
-      fmt.addStyleName(1, 1, Gerrit.RESOURCES.css().dataHeader());
-      fmt.addStyleName(1, 2, Gerrit.RESOURCES.css().dataHeader());
-    }
-
-    void deleteChecked() {
-      final Set<AccountProjectWatch.Key> ids = getCheckedIds();
-      if (!ids.isEmpty()) {
-        Util.ACCOUNT_SVC.deleteProjectWatches(ids,
-            new GerritCallback<VoidResult>() {
-              public void onSuccess(final VoidResult result) {
-                remove(ids);
-              }
-            });
-      }
-    }
-
-    void remove(Set<AccountProjectWatch.Key> ids) {
-      for (int row = 1; row < table.getRowCount();) {
-        final AccountProjectWatchInfo k = getRowItem(row);
-        if (k != null && ids.contains(k.getWatch().getKey())) {
-          table.removeRow(row);
-        } else {
-          row++;
-        }
-      }
-    }
-
-    Set<AccountProjectWatch.Key> getCheckedIds() {
-      final Set<AccountProjectWatch.Key> ids =
-          new HashSet<AccountProjectWatch.Key>();
-      for (int row = 1; row < table.getRowCount(); row++) {
-        final AccountProjectWatchInfo k = getRowItem(row);
-        if (k != null && ((CheckBox) table.getWidget(row, 1)).getValue()) {
-          ids.add(k.getWatch().getKey());
-        }
-      }
-      return ids;
-    }
-
-    void insertWatch(final AccountProjectWatchInfo k) {
-      final String newName = k.getProject().getName();
-      int row = 1;
-      for (; row < table.getRowCount(); row++) {
-        final AccountProjectWatchInfo i = getRowItem(row);
-        if (i != null && i.getProject().getName().compareTo(newName) >= 0) {
-          break;
-        }
-      }
-
-      table.insertRow(row);
-      applyDataRowStyle(row);
-      populate(row, k);
-    }
-
-    void display(final List<AccountProjectWatchInfo> result) {
-      while (2 < table.getRowCount())
-        table.removeRow(table.getRowCount() - 1);
-
-      for (final AccountProjectWatchInfo k : result) {
-        final int row = table.getRowCount();
-        table.insertRow(row);
-        applyDataRowStyle(row);
-        populate(row, k);
-      }
-    }
-
-    void populate(final int row, final AccountProjectWatchInfo info) {
-      final FlowPanel fp = new FlowPanel();
-      fp.add(new ProjectLink(info.getProject().getNameKey(), Status.NEW));
-      if (info.getWatch().getFilter() != null) {
-        Label filter = new Label(info.getWatch().getFilter());
-        filter.setStyleName(Gerrit.RESOURCES.css().watchedProjectFilter());
-        fp.add(filter);
-      }
-
-      table.setWidget(row, 1, new CheckBox());
-      table.setWidget(row, 2, fp);
-
-      addNotifyButton(AccountProjectWatch.Type.NEW_CHANGES, info, row, 3);
-      addNotifyButton(AccountProjectWatch.Type.COMMENTS, info, row, 4);
-      addNotifyButton(AccountProjectWatch.Type.SUBMITS, info, row, 5);
-
-      final FlexCellFormatter fmt = table.getFlexCellFormatter();
-      fmt.addStyleName(row, 1, Gerrit.RESOURCES.css().iconCell());
-      fmt.addStyleName(row, 2, Gerrit.RESOURCES.css().dataCell());
-      fmt.addStyleName(row, 3, Gerrit.RESOURCES.css().dataCell());
-      fmt.addStyleName(row, 4, Gerrit.RESOURCES.css().dataCell());
-      fmt.addStyleName(row, 5, Gerrit.RESOURCES.css().dataCell());
-
-      setRowItem(row, info);
-    }
-
-    protected void addNotifyButton(final AccountProjectWatch.Type type,
-        final AccountProjectWatchInfo info, final int row, final int col) {
-      final CheckBox cbox = new CheckBox();
-
-      cbox.addClickHandler(new ClickHandler() {
-        @Override
-        public void onClick(final ClickEvent event) {
-          final boolean oldVal = info.getWatch().isNotify(type);
-          info.getWatch().setNotify(type, cbox.getValue());
-          Util.ACCOUNT_SVC.updateProjectWatch(info.getWatch(),
-              new GerritCallback<VoidResult>() {
-                public void onSuccess(final VoidResult result) {
-                }
-
-                @Override
-                public void onFailure(final Throwable caught) {
-                  info.getWatch().setNotify(type, oldVal);
-                  cbox.setValue(oldVal);
-                  super.onFailure(caught);
-                }
-              });
-        }
-      });
-
-      cbox.setValue(info.getWatch().isNotify(type));
-      table.setWidget(row, col, cbox);
-    }
-  }
 }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java
new file mode 100644
index 0000000..be808fa
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java
@@ -0,0 +1,178 @@
+// Copyright (C) 2010 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.gerrit.client.account;
+
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.client.ui.FancyFlexTable;
+import com.google.gerrit.client.ui.ProjectLink;
+import com.google.gerrit.common.data.AccountProjectWatchInfo;
+import com.google.gerrit.reviewdb.AccountProjectWatch;
+import com.google.gerrit.reviewdb.Change.Status;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.user.client.DOM;
+import com.google.gwt.user.client.ui.CheckBox;
+import com.google.gwt.user.client.ui.FlowPanel;
+import com.google.gwt.user.client.ui.Label;
+import com.google.gwt.user.client.ui.FlexTable.FlexCellFormatter;
+import com.google.gwtjsonrpc.client.VoidResult;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class MyWatchesTable extends FancyFlexTable<AccountProjectWatchInfo> {
+
+  public MyWatchesTable() {
+    table.setWidth("");
+    table.insertRow(1);
+    table.setText(0, 2, Util.C.watchedProjectName());
+    table.setText(0, 3, Util.C.watchedProjectColumnEmailNotifications());
+
+    final FlexCellFormatter fmt = table.getFlexCellFormatter();
+    fmt.addStyleName(0, 1, Gerrit.RESOURCES.css().iconHeader());
+    fmt.addStyleName(0, 2, Gerrit.RESOURCES.css().dataHeader());
+    fmt.addStyleName(0, 3, Gerrit.RESOURCES.css().dataHeader());
+    fmt.setRowSpan(0, 0, 2);
+    fmt.setRowSpan(0, 1, 2);
+    fmt.setRowSpan(0, 2, 2);
+    DOM.setElementProperty(fmt.getElement(0, 3), "align", "center");
+
+    fmt.setColSpan(0, 3, 3);
+    table.setText(1, 0, Util.C.watchedProjectColumnNewChanges());
+    table.setText(1, 1, Util.C.watchedProjectColumnAllComments());
+    table.setText(1, 2, Util.C.watchedProjectColumnSubmittedChanges());
+    fmt.addStyleName(1, 0, Gerrit.RESOURCES.css().dataHeader());
+    fmt.addStyleName(1, 1, Gerrit.RESOURCES.css().dataHeader());
+    fmt.addStyleName(1, 2, Gerrit.RESOURCES.css().dataHeader());
+  }
+
+  public void deleteChecked() {
+    final Set<AccountProjectWatch.Key> ids = getCheckedIds();
+    if (!ids.isEmpty()) {
+      Util.ACCOUNT_SVC.deleteProjectWatches(ids,
+          new GerritCallback<VoidResult>() {
+            public void onSuccess(final VoidResult result) {
+              remove(ids);
+            }
+          });
+    }
+  }
+
+  protected void remove(Set<AccountProjectWatch.Key> ids) {
+    for (int row = 1; row < table.getRowCount();) {
+      final AccountProjectWatchInfo k = getRowItem(row);
+      if (k != null && ids.contains(k.getWatch().getKey())) {
+        table.removeRow(row);
+      } else {
+        row++;
+      }
+    }
+  }
+
+  protected Set<AccountProjectWatch.Key> getCheckedIds() {
+    final Set<AccountProjectWatch.Key> ids =
+        new HashSet<AccountProjectWatch.Key>();
+    for (int row = 1; row < table.getRowCount(); row++) {
+      final AccountProjectWatchInfo k = getRowItem(row);
+      if (k != null && ((CheckBox) table.getWidget(row, 1)).getValue()) {
+        ids.add(k.getWatch().getKey());
+      }
+    }
+    return ids;
+  }
+
+  public void insertWatch(final AccountProjectWatchInfo k) {
+    final String newName = k.getProject().getName();
+    int row = 1;
+    for (; row < table.getRowCount(); row++) {
+      final AccountProjectWatchInfo i = getRowItem(row);
+      if (i != null && i.getProject().getName().compareTo(newName) >= 0) {
+        break;
+      }
+    }
+
+    table.insertRow(row);
+    applyDataRowStyle(row);
+    populate(row, k);
+  }
+
+  public void display(final List<AccountProjectWatchInfo> result) {
+    while (2 < table.getRowCount())
+      table.removeRow(table.getRowCount() - 1);
+
+    for (final AccountProjectWatchInfo k : result) {
+      final int row = table.getRowCount();
+      table.insertRow(row);
+      applyDataRowStyle(row);
+      populate(row, k);
+    }
+  }
+
+  protected void populate(final int row, final AccountProjectWatchInfo info) {
+    final FlowPanel fp = new FlowPanel();
+    fp.add(new ProjectLink(info.getProject().getNameKey(), Status.NEW));
+    if (info.getWatch().getFilter() != null) {
+      Label filter = new Label(info.getWatch().getFilter());
+      filter.setStyleName(Gerrit.RESOURCES.css().watchedProjectFilter());
+      fp.add(filter);
+    }
+
+    table.setWidget(row, 1, new CheckBox());
+    table.setWidget(row, 2, fp);
+
+    addNotifyButton(AccountProjectWatch.Type.NEW_CHANGES, info, row, 3);
+    addNotifyButton(AccountProjectWatch.Type.COMMENTS, info, row, 4);
+    addNotifyButton(AccountProjectWatch.Type.SUBMITS, info, row, 5);
+
+    final FlexCellFormatter fmt = table.getFlexCellFormatter();
+    fmt.addStyleName(row, 1, Gerrit.RESOURCES.css().iconCell());
+    fmt.addStyleName(row, 2, Gerrit.RESOURCES.css().dataCell());
+    fmt.addStyleName(row, 3, Gerrit.RESOURCES.css().dataCell());
+    fmt.addStyleName(row, 4, Gerrit.RESOURCES.css().dataCell());
+    fmt.addStyleName(row, 5, Gerrit.RESOURCES.css().dataCell());
+
+    setRowItem(row, info);
+  }
+
+  protected void addNotifyButton(final AccountProjectWatch.Type type,
+      final AccountProjectWatchInfo info, final int row, final int col) {
+    final CheckBox cbox = new CheckBox();
+
+    cbox.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        final boolean oldVal = info.getWatch().isNotify(type);
+        info.getWatch().setNotify(type, cbox.getValue());
+        Util.ACCOUNT_SVC.updateProjectWatch(info.getWatch(),
+            new GerritCallback<VoidResult>() {
+              public void onSuccess(final VoidResult result) {
+              }
+
+              @Override
+              public void onFailure(final Throwable caught) {
+                info.getWatch().setNotify(type, oldVal);
+                cbox.setValue(oldVal);
+                super.onFailure(caught);
+              }
+            });
+      }
+    });
+
+    cbox.setValue(info.getWatch().isNotify(type));
+    table.setWidget(row, col, cbox);
+  }
+}
-- 
1.7.2.2

